### PR TITLE
caja-directory-async: trustred launchers

### PR DESCRIPTION
**Pull request by rbuj**
See #1466

### CODE REVIEW

**Comment by lukefromdc** [comment]
Is this intended to let users manually mark a .desktop file as trusted after examination or not?

**Comment by rbuj** [comment]
I haven't decided this part yet, but we could store a new gvfs metadata (.local/share/gvfs-metadata) such as trusted-launcher="yes"

**Comment by lukefromdc** [comment]
I am opposed to force-disabling user-created .desktop files with no option to override. In fact, I use them myself

**Comment by rbuj** [comment]
@lukefromdc done

**Comment by HacKurx** [comment]
For me, only the .desktop files coming from /usr/share/applications/ are trusted (because it belongs to root).
The goal is not to prevent the user from creating *.desktop files just distinguish them visually.

There are many different attacks with *.desktop files. Here is another one:

Imagine that Bob forgets to log out. Someone passes and changes the firefox shortcut in mate-panel or on the desktop to replace the legitimate command with a malicious one.
Bob comes back and clicks it because for him it's still firefox (same icon, no visual warning that the shortcut has changed).

Ideally, the file '/home/Bob/Desktop/firefox.desktop' should look the same only if it is identical (sha256) to the file '/usr/share/applications/firefox.desktop'. If the file '/home/Bob/Desktop/firefox.desktop' is modified or created by the user then it should have a visual emblem imposed.

**Comment by rbuj** [comment]
The alert emblem can also be replaced with a transparent image by changing the icon theme, which may be located in user home.

**Comment by HacKurx** [comment]
So for the more cautious among us, we need a new option in caja :

@artebin
> It is also critical to have an option in the Preferences of Caja: Always show complete file name in file explorer window.

Thanks.

**Comment by artebin** [comment]
Or maybe the other way around: an option FALSE by default named "Hide .desktop extensions in Desktop folder only"

**Review by lukefromdc — COMMENTED** [review]
Here's what I got on a test build and install (which went fine, and no surprises elsewhere at runtime):

Existing user-made desktop file on desktop no change in behavior
Moved that file to another directory, and the "launch anyway/mark as trusted/cancel dialog came up and worked as expected. 

NEW (made in same run of caja) user-made desktop file on desktop no change in behavior

Moved that file to another directory, and the "launch anyway/mark as trusted/cancel dialog comes up, but marking as trusted did not work until caja is restarted. When caja was restarted, the launcher was now trusted but not before. Selecting "launch anyway" always worked as expected

This was verified by testing it twice, after caja restarts

**Comment by lukefromdc** [comment]
Does anyone else want to review this (it is a behavior change) before merging it? I forgot to re-select coreteam for another review

**Comment by raveit65** [comment]
LiveCDs are coming with desktop files like `Install-Live-CD-to-disc.`
So this PR will mark them as un-trusted which isn't true.
I am not sure if we really should do this.

**Comment by rbuj** [comment]
> LiveCDs are coming with desktop files like `Install-Live-CD-to-disc.`
> So this PR will mark them as un-trusted which isn't true.

They are marked as trusted in `caja-directory-async: trust also launchers from user desktop`.

**Comment by lukefromdc** [comment]
We cannot break liveCD environments using MATE. That could kill things like Ubuntu-MATE outright

File: libcaja-private/caja-file-utilities.c [hunk 1]
```diff
@@ -1134,6 +1134,33 @@ caja_is_in_system_dir (GFile *file)
     return res;
 }
 
+gboolean
+caja_is_in_desktop_dir (GFile *file)
+{
+    const char *desktop_dir;
```
Comment (raveit65) [inline libcaja-private/caja-file-utilities.c:1140]: Build warning.
```
caja-file-utilities.c:1140:17: warning: declaration shadows a variable in the global scope [-Wshadow]

    const char *desktop_dir;
```
see https://travis-ci.org/github/mate-desktop/caja/jobs/745617527#L4604

**Comment by raveit65** [comment]
> > LiveCDs are coming with desktop files like `Install-Live-CD-to-disc.`
> > So this PR will mark them as un-trusted which isn't true.
> 
> They are marked as trusted in `caja-directory-async: trust also launchers from user desktop`.

Yes. now i see it and it works like expected.
Desktop files on the desktop are marked as trusted but not desktop files in eg. home dir of user account.

File: libcaja-private/caja-file-utilities.c (hunk 1 above)
Comment (rbuj) [reply libcaja-private/caja-file-utilities.c]: @raveit65 done

**Review by raveit65 — APPROVED** [review]
LGTM,
does what it promise.

**Review by vkareh — APPROVED** [review]
This is a nice feature.

LGTM

**Pull request merged**